tools/tcpcong.py

```diff
@@ -0,0 +1,484 @@
+    u32 daddr;
+    u16 lport;
+    u16 dport;
+}ipv4_flow_key_t;
+
+typedef struct ipv6_flow_key {
+    unsigned __int128 saddr;
+    unsigned __int128 daddr;
+    u16 lport;
+    u16 dport;
+}ipv6_flow_key_t;
+
+typedef struct process_key {
+    char comm[TASK_COMM_LEN];
+    u32  pid;
+    u32  tid;
+}process_key_t;
+
+typedef struct ipv4_flow_val {
+    ipv4_flow_key_t ipv4_key;
+    u16  cong_state;
+}ipv4_flow_val_t;
+
+typedef struct ipv6_flow_val {
+    ipv6_flow_key_t ipv6_key;
+    u16  cong_state;
+}ipv6_flow_val_t;
+
+
+BPF_HASH(start_ipv4, process_key_t, ipv4_flow_val_t);
+BPF_HASH(start_ipv6, process_key_t, ipv6_flow_val_t);
+
+typedef struct data_val {
+    DEF_TEXT
+    u64  last_ts;
+    u16  last_cong_stat;
+}data_val_t;
+
+BPF_HASH(ipv4_stat, ipv4_flow_key_t, data_val_t);
+BPF_HASH(ipv6_stat, ipv6_flow_key_t, data_val_t);
+
+HIST_TABLE
+
+int entry_state_update_func(struct pt_regs *ctx, struct sock *sk)
+{
+    u32 tid = bpf_get_current_pid_tgid();
+    u32 pid = (bpf_get_current_pid_tgid() >> 32);
+    process_key_t key = {};
+    bpf_get_current_comm(&key.comm, sizeof(key.comm));
+    key.pid = pid;
+    key.tid = tid;
+    u16 dport = 0, lport = 0;
+    u64 family = sk->__sk_common.skc_family;
+    struct inet_connection_sock *icsk = inet_csk(sk);
+    struct cong {
+        u8  cong_stat:5,
+            ca_inited:1,
+            ca_setsockopt:1,
+            ca_dstlocked:1;
+    }cong_status;
+
+    bpf_probe_read_kernel(&cong_status, sizeof(cong_status),
+        (void *)((long)&icsk->icsk_retransmits) - 1);
+    if (family == AF_INET) {
+        ipv4_flow_val_t ipv4_val = {0};
+        ipv4_val.ipv4_key.saddr = sk->__sk_common.skc_rcv_saddr;
+        ipv4_val.ipv4_key.daddr = sk->__sk_common.skc_daddr;
+        ipv4_val.ipv4_key.lport = sk->__sk_common.skc_num;
+        dport = sk->__sk_common.skc_dport;
+        dport = ntohs(dport);
+        lport = ipv4_val.ipv4_key.lport;
+        FILTER_LPORT
+        FILTER_DPORT
+        ipv4_val.ipv4_key.dport = dport;
+        ipv4_val.cong_state = cong_status.cong_stat + 1;
+        start_ipv4.update(&key, &ipv4_val);
+    } else if (family == AF_INET6) {
+        ipv6_flow_val_t ipv6_val = {0};
+        bpf_probe_read_kernel(&ipv6_val.ipv6_key.saddr,
+            sizeof(ipv6_val.ipv6_key.saddr),
+            &sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+        bpf_probe_read_kernel(&ipv6_val.ipv6_key.daddr,
+            sizeof(ipv6_val.ipv6_key.daddr),
+            &sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
+        ipv6_val.ipv6_key.lport = sk->__sk_common.skc_num;
+        dport = sk->__sk_common.skc_dport;
+        dport = ntohs(dport);
+        lport = ipv6_val.ipv6_key.lport;
+        FILTER_LPORT
+        FILTER_DPORT
+        ipv6_val.ipv6_key.dport = ntohs(dport);
+        ipv6_val.cong_state = cong_status.cong_stat + 1;
+        start_ipv6.update(&key, &ipv6_val);
+    }
+    return 0;
+}
+
+int ret_state_update_func(struct pt_regs *ctx, struct sock *sk)
```

Comment:
This won't work, for return probe, we don't have 'struct sock *sk' available. To get 'struct sock *sk', you need to save it in the hash table with key tid or task ptr, and retrieve it in the kretprobe.

Do we have potential for nested bpf program execution here, one is in process context and another in softirq context? In such nested case, the bpf program in softirq context will not get executed due to kernel internal checking, so we could miss some events.

Alternatively, you can kfunc/kretfunc which does give you function parameters in kretfunc. It is only available recent kernels though.

Reply:
okay, should store the paremeter 'struct sock *sk' in a hash table.
yes, your worry is reasonable, if there is  nested execution between softirq context  and process context ,  we should need another  context  type(softirq, process) to record , then can cover this case.  does bcc have a function to get the context type? if so, please tell me and i will add code for this case.  thanks very much!

Reply:
`context type(softirq, process) ` won't work since this is kprobe and in the kernel (kernel/trace/bpf_trace.c) we have
```
        if (unlikely(__this_cpu_inc_return(bpf_prog_active) != 1)) {
                /*
                 * since some bpf program is already running on this cpu,
                 * don't call into another bpf program (same or different)
                 * and don't send kprobe event into ring-buffer,
                 * so return zero here
                 */
                ret = 0;
                goto out;
        }
```
Basically nested kprobe programs not allowed.

kfunc approach is better and it allows nested bpf programs as long as they are different programs. But kfunc will not work for earlier kernels.

Maybe putting a warning/caveat in the beginning of tools is okay.

Reply:
@yonghong-song , thanks for your comments. okay, i will add the kfunc/kretfunc for the tools when it's available.